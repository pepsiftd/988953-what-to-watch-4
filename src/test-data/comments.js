const reviews = [
  {
    id: 1,
    user: {
      id: 4,
      name: `Courtney Cox`,
    },
    rating: 5.1,
    comment: `I don't know what to say but I can't help posting`,
    date: new Date(`August 3, 2020 UTC`),
  },
  {
    id: 2,
    user: {
      id: 2,
      name: `Mockingbird`,
    },
    rating: 8.3,
    comment: `This is not a movie of a simple kind`,
    date: new Date(`July 15, 2020 UTC`),
  },
  {
    id: 3,
    user: {
      id: 154,
      name: `Pelagia`,
    },
    rating: 1.5,
    comment: `Super mega exquisitely boring`,
    date: new Date(`June 1, 2020 UTC`),
  }
];

export {reviews};
