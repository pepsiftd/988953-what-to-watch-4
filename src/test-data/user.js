const userInfo = {
  id: 0,
  email: `example@email.com`,
  name: `Sarah Connor`,
  avatar: `img/avatar.jpg`,
};

export {userInfo};
