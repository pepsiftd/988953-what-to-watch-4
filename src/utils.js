const getUniqueItems = (list) => {
  return list.filter((it, i, array) => {
    return array.indexOf(it) === i;
  });
};

const extend = (a, b) => {
  return Object.assign({}, a, b);
};

const getMoviesByGenre = (movies, genre) => {
  return genre === `All genres` ? movies : movies.filter((movie) => movie.genre === genre);
};

const getGenresFromMovies = (movies) => {
  return getUniqueItems(movies.map((movie) => movie.genre));
};

export {extend, getMoviesByGenre, getUniqueItems, getGenresFromMovies};
