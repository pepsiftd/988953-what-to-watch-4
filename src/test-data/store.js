import {NameSpace} from '@/reducer/name-space';
import {movies, promoMovie} from '@/test-data/movies';
import {userInfo} from '@/test-data/user';
import {reviews} from '@/test-data/comments';
import {ALL_GENRES_FILTER} from '@/const';

const store = {
  [NameSpace.DATA]: {
    movies,
    promoMovie,
    favoriteMovies: [],
    reviews,
  },
  [NameSpace.APP]: {
    currentGenre: ALL_GENRES_FILTER,
  },
  [NameSpace.USER]: {
    authorizationStatus: `AUTHORIZED`,
    authorizationError: 200,
    authorizationInfo: userInfo,
  },
};

export {store};
