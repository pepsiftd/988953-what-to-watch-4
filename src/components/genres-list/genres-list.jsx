import React from 'react';
import PropTypes from 'prop-types';

import {GenresListItem} from '@/components/genres-list-item/genres-list-item';
import {getGenresFromMovies} from '@/utils';
import {MAX_GENRES_AMOUNT, ALL_GENRES_FILTER} from '@/const';

const GenresList = ({movies, renderItem, setActiveItem, onTitleClick}) => {
  const genres = [ALL_GENRES_FILTER, ...getGenresFromMovies(movies).slice(0, MAX_GENRES_AMOUNT)];

  return (
    <ul className="catalog__genres-list">
      {genres.map((genre) => {
        return (
          renderItem(
              GenresListItem,
              genre,
              {
                key: genre,
                genre,
                onTitleClick: (...args) => {
                  setActiveItem(...args);
                  onTitleClick(...args);
                },
              }
          )
        );
      })}

    </ul>
  );
};

GenresList.propTypes = {
  movies: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    genre: PropTypes.string.isRequired,
  }).isRequired).isRequired,
  renderItem: PropTypes.func.isRequired,
  setActiveItem: PropTypes.func.isRequired,
  onTitleClick: PropTypes.func.isRequired,
};

export {GenresList};
