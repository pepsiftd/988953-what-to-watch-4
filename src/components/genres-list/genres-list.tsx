import * as React from 'react';

import {GenresListItem} from '@/components/genres-list-item/genres-list-item';
import {getGenresFromMovies} from '@/utils';
import {MAX_GENRES_AMOUNT, ALL_GENRES_FILTER} from '@/const';

interface Props {
  movies: {
    id: number;
    genre: string;
  }[];
  renderItem: (ItemComponent: React.FunctionComponent, itemId: number | string, itemProps: Record<string, unknown>) => React.ReactNode;
  setActiveItem: (id: number | string) => void;
  onTitleClick: (id: number | string) => void;
}

const GenresList: React.FunctionComponent<Props> = ({movies, renderItem, setActiveItem, onTitleClick}) => {
  const genres = [ALL_GENRES_FILTER, ...getGenresFromMovies(movies).slice(0, MAX_GENRES_AMOUNT)];

  return (
    <ul className="catalog__genres-list">
      {genres.map((genre) => {
        return (
          renderItem(
              GenresListItem,
              genre,
              {
                key: genre,
                genre,
                onTitleClick: (args) => {
                  setActiveItem(args);
                  onTitleClick(args);
                },
              }
          )
        );
      })}

    </ul>
  );
};

export {GenresList};
