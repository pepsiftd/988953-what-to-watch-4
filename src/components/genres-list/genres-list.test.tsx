import React from 'react';
import renderer from 'react-test-renderer';

import {noop} from '@/utils';
import {GenresList} from './genres-list';
import {movies} from '@/test-data/movies';

it(`GenresList renders correctly`, () => {
  const tree = renderer.create(
      <GenresList
        movies={movies}
        renderItem={noop}
        setActiveItem={noop}
        onTitleClick={noop}
      />
  );

  expect(tree).toMatchSnapshot();
});
