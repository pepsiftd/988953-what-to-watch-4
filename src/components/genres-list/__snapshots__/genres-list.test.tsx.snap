// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenresList renders correctly 1`] = `
<ul
  className="catalog__genres-list"
/>
`;
